flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java

```diff
@@ -0,0 +1,212 @@
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequest;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequestHandler;
+import org.apache.flink.runtime.operators.coordination.CoordinationResponse;
+import org.apache.flink.runtime.operators.coordination.OperatorCoordinator;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * {@link OperatorCoordinator} for {@link CollectSinkFunction}.
+ *
+ * <p>This coordinator only forwards requests and responses from clients and sinks,
+ * it does not store any state in itself.
+ */
+public class CollectSinkOperatorCoordinator implements OperatorCoordinator, CoordinationRequestHandler {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CollectSinkOperatorCoordinator.class);
+
+	private InetSocketAddress address;
+	private Socket socket;
+
+	private ExecutorService executorService;
+
+	@Override
+	public void start() throws Exception {
+		this.executorService = Executors.newFixedThreadPool(1);
+	}
+
+	@Override
+	public void close() throws Exception {
+		closeCurrentConnection();
+		this.executorService.shutdown();
+	}
+
+	@Override
+	public void handleEventFromOperator(int subtask, OperatorEvent event) throws Exception {
+		Preconditions.checkArgument(
+			event instanceof CollectSinkAddressEvent, "Operator event must be a CollectSinkAddressEvent");
+		address = ((CollectSinkAddressEvent) event).getAddress();
+		LOG.debug("Received sink socket server address: " + address);
+
+		// this event is sent when the sink function starts, so we remove the old socket if it is present
+		socket = null;
+	}
+
+	@Override
+	public CompletableFuture<CoordinationResponse> handleCoordinationRequest(CoordinationRequest request) {
+		Preconditions.checkArgument(
+			request instanceof CollectCoordinationRequest,
+			"Coordination request must be a CollectCoordinationRequest");
+
+		CompletableFuture<CoordinationResponse> responseFuture = new CompletableFuture<>();
+		executorService.submit(() -> handleRequestImpl((CollectCoordinationRequest) request, responseFuture));
+		return responseFuture;
+	}
+
+	private void handleRequestImpl(
+			CollectCoordinationRequest request,
+			CompletableFuture<CoordinationResponse> responseFuture) {
+		// there is no thread safety issue in this method
```

Comment:
not really, event handling is executed by JM's rpc thread.

Reply:
We have only 1 thread in the thread pool. This method will not be executed concurrently. In current implementation we also expect only 1 client to read results from this coordinator.

Reply:
I was talking about `handleEventFromOperator` method, which isn't triggered by client.

Reply:
Yes. But this `handleRequestImpl` is executed by the thread pool, not by the JM's rpc thread. `handleEventFromOperator` only submits this method to the thread pool and returns a `CompletableFuture`.